src/access/RequiredPrivilege.cpp

```diff
@@ -0,0 +1,214 @@
+namespace {
+
+using namespace chip;
+using namespace chip::Access;
+
+// Privilege override entries are stored in a table per operation (read attribute,
+// write attribute, invoke command, read entry). Entries in each table are sorted
+// by their primary compound key of cluster/endpoint/field, and therefore cannot
+// have duplicate privilege for any such combination. Cluster cannot be invalid,
+// so that is used as a sentinel for empty/unused entries, which are kept at the
+// end of the table. Endpoint and field can be invalid, which means wildcard.
+// Sorting the entries in this way allows a cluster to override specific endpoints
+// or fields with privilege, while still using wildcard to override all the
+// unspecified ones.
+struct PrivilegeOverride
+{
+    ClusterId mCluster;
+    EndpointId mEndpoint;
+    FieldId mField;
+    Privilege mPrivilege;
+
+    PrivilegeOverride() : mCluster(kInvalidClusterId), mEndpoint(kInvalidEndpointId), mField(kInvalidFieldId) {}
+
+    PrivilegeOverride(ClusterId cluster, EndpointId endpoint, FieldId field, Privilege privilege) :
+        mCluster(cluster), mEndpoint(endpoint), mField(field), mPrivilege(privilege)
+    {}
+
+    static_assert(sizeof(FieldId) >= sizeof(AttributeId), "FieldId must be able to hold AttributeId");
+    static_assert(sizeof(FieldId) >= sizeof(CommandId), "FieldId must be able to hold CommandId");
+    static_assert(sizeof(FieldId) >= sizeof(EventId), "FieldId must be able to hold EventId");
+};
+
+PrivilegeOverride
+    privilegeOverrideForReadAttribute[CHIP_CONFIG_ACCESS_CONTROL_MAX_REQUIRED_PRIVILEGE_OVERRIDES_FOR_READ_ATTRIBUTE] = {
+        PrivilegeOverride(0x0000'001f, kInvalidEndpointId, kInvalidFieldId, Privilege::kAdminister),
+    };
+
+PrivilegeOverride
+    privilegeOverrideForWriteAttribute[CHIP_CONFIG_ACCESS_CONTROL_MAX_REQUIRED_PRIVILEGE_OVERRIDES_FOR_WRITE_ATTRIBUTE] = {
+        PrivilegeOverride(0x0000'001f, kInvalidEndpointId, kInvalidFieldId, Privilege::kAdminister),
+    };
+
+PrivilegeOverride
+    privilegeOverrideForInvokeCommand[CHIP_CONFIG_ACCESS_CONTROL_MAX_REQUIRED_PRIVILEGE_OVERRIDES_FOR_INVOKE_COMMAND] = {
+        PrivilegeOverride(0x0000'001f, kInvalidEndpointId, kInvalidFieldId, Privilege::kAdminister),
+    };
+
+PrivilegeOverride privilegeOverrideForReadEvent[CHIP_CONFIG_ACCESS_CONTROL_MAX_REQUIRED_PRIVILEGE_OVERRIDES_FOR_READ_EVENT] = {
+    PrivilegeOverride(0x0000'001f, kInvalidEndpointId, kInvalidFieldId, Privilege::kAdminister),
+};
+
+enum class Operation
+{
+    kReadAttribute  = 0,
+    kWriteAttribute = 1,
+    kInvokeCommand  = 2,
+    kReadEvent      = 3
+};
+
+PrivilegeOverride * const privilegeOverride[] = { privilegeOverrideForReadAttribute, privilegeOverrideForWriteAttribute,
+                                                  privilegeOverrideForInvokeCommand, privilegeOverrideForReadEvent };
+
+const size_t maxPrivilegeOverride[] = { sizeof(privilegeOverrideForReadAttribute) / sizeof(privilegeOverrideForReadAttribute[0]),
```

Comment:
Please use our existing `ArraySize()` thing, which is much clearer as to what's being computed.

Reply:
Are those array static? why we still need to compute the array size instead of just using CHIP_CONFIG_ACCESS_CONTROL_MAX_REQUIRED_PRIVILEGE_OVERRIDES_FOR_WRITE_ATTRIBUTE?

Reply:
Ah, yes I will.

Reply:
Look at how the four operations funnel through one data driven function.

Reply:
Using ArraySize, made the arrays const, removed the config flags.